### PR TITLE
♻️  Update Keyhole trait to use shared values and apply transitions

### PR DESCRIPTION
**Pull request by andretortolano**
this is another piece of the puzzle, for the new 2.0 trait system in place.

Keyhole trait gets information from target-rectangle and the animation trait to transition between rectangle/circular shapes and position on the screen.

https://user-images.githubusercontent.com/5244805/207149108-2f516be4-28a4-4ac3-a973-56ed64fe731c.mov

### CODE REVIEW

File: appcues/src/main/java/com/appcues/trait/TraitKoin.kt [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
 import com.appcues.trait.appcues.SkippableTrait
 import com.appcues.trait.appcues.StepAnimationTrait
 import com.appcues.trait.appcues.StickyContentTrait
+import com.appcues.trait.appcues.TargetElementTrait
+import com.appcues.trait.appcues.TargetRectangleTrait
 import org.koin.dsl.ScopeDSL
 
 internal object TraitKoin : KoinScopePlugin {
 
+    @Suppress("LongMethod")
```
Comment (iujames) [inline appcues/src/main/java/com/appcues/trait/TraitKoin.kt:19]: does it make sense to suppress at config level for any `*Koin` files that are just doing these scope declarations?

File: appcues/src/main/java/com/appcues/trait/appcues/BackdropKeyholeTrait.kt [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    override fun BoxScope.BackdropDecorate(content: @Composable BoxScope.() -> Unit) {
+        val density = LocalDensity.current
+        val metadata = LocalAppcuesStepMetadata.current
+
+        val actualRect = remember(metadata) {
+            (metadata.actual[TargetElementTrait.METADATA_TARGET_RECT] as TargetRect?) ?: emptyRect
+        }
+
+        val duration = remember(metadata) {
+            (metadata.actual[StepAnimationTrait.METADATA_ANIMATION_DURATION] as Int?) ?: DEFAULT_ANIMATION
```
Comment (iujames) [inline appcues/src/main/java/com/appcues/trait/appcues/BackdropKeyholeTrait.kt:76]: if no animation trait is present, should we have no animation or some default animation? cc @mmaatttt to keep in sync

File: appcues/src/main/java/com/appcues/trait/appcues/BackdropKeyholeTrait.kt [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        )
+        val width = animateFloatAsState(if (shape == RECTANGLE) actualRect.width.toFloat() else circleSize, animationSpec = animation)
+        val height = animateFloatAsState(if (shape == RECTANGLE) actualRect.height.toFloat() else circleSize, animationSpec = animation)
+        val cornerRadius =
+            animateFloatAsState(getCornerRadius(actualRect.width.toFloat(), actualRect.height.toFloat()), animationSpec = animation)
+
+        Box(
+            modifier = Modifier
+                .fillMaxSize()
+                // set background color
```
Comment (iujames) [inline appcues/src/main/java/com/appcues/trait/appcues/BackdropKeyholeTrait.kt:108]: maybe a copy/paste comment?

File: appcues/src/main/java/com/appcues/trait/appcues/TargetElementTrait.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+        const val METADATA_TARGET_RECT = "targetRectangle"
+    }
+
+    data class TargetRect(
+        val x: Int,
+        val y: Int,
+        val width: Int,
+        val height: Int,
+    )
```
Comment (iujames) [inline appcues/src/main/java/com/appcues/trait/appcues/TargetElementTrait.kt:23]: not critical, but can we use something built in like `android.graphics.Rect`?

**Review by iujames — APPROVED** [review]
rollin along! would nice to copy your demo video here for historical reference on the PR

File: appcues/src/main/java/com/appcues/trait/appcues/BackdropKeyholeTrait.kt (hunk 2 above)
Comment (andretortolano) [reply appcues/src/main/java/com/appcues/trait/appcues/BackdropKeyholeTrait.kt:76]: going for 350millis as the default but open for suggestions

**Pull request merged**